Dockerfile.mariadb - use included healthcheck

grooverdan opened the pull request:
MARIADB_MYSQL_LOCALHOST_USER=1 creates the mysql@localhost user that the healthcheck uses with --su=mysql.

`--innodb_initialized` was chosen as a fairly basic test to ensure that any crash recovery is complete.

grooverdan committed:
Dockerfile.mariadb - use included healthcheck

MARIADB_MYSQL_LOCALHOST_USER=1 creates the mysql@localhost user that the healthcheck uses with --su=mysql.

`--innodb_initialized` was chosen as a fairly basic test to ensure that any crash recovery is complete.

tianon commented on mysql/Dockerfile.mariadb:
```diff
@@ -1,5 +1,5 @@
+ENV MARIADB_MYSQL_LOCALHOST_USER=1
 
-HEALTHCHECK CMD ["docker-healthcheck"]
+HEALTHCHECK CMD ["/usr/local/bin/healthcheck.sh", "--su=mysql", "--innodb_initialized"]
```
```suggestion
HEALTHCHECK CMD ["healthcheck.sh", "--su=mysql", "--innodb_initialized"]
```

Does it always use InnoDB by default?  (Will this check still be meaningful/useful if InnoDB _isn't_ used?)

grooverdan replied:
Technically I can't start without innodb at the moment;
```
$ podman run --rm -e MARIADB_ROOT_PASSWORD=bob  mariadb:10.5 --skip-innodb
2022-03-09 22:42:06+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.15+maria~focal started.
2022-03-09 22:42:07+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2022-03-09 22:42:07+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.5.15+maria~focal started.
2022-03-09 22:42:07+00:00 [Note] [Entrypoint]: Initializing database files
2022-03-09 22:42:07 0 [ERROR] Unknown/unsupported storage engine: InnoDB
2022-03-09 22:42:07 0 [ERROR] Aborting
```
If `--innodb_initialized` wasn't there another test from the script would need to be used.

tianon replied:
Fair -- I guess my question was more "is there a more generic test we could use" but it sounds like this is probably the best one.

My next concern is that the script appears to use different `mysql` options depending on the test, so some might use the unix socket, right?  So this will potentially be reporting the container as "healthy" even while the initialization scripts are still running, especially if they're taking a long time (like a really large import)?

Should we always start with `--connect` (that uses `--protocol tcp` explicitly, presumably to combat this issue)?

grooverdan replied:
You are right in that `innodb_initialized` and `mariadbupgrade` are the ones that could report healthy on their own ahead of fully started. I'll get that documented.

grooverdan committed:
Update Dockerfile.mariadb

with connect so ensure we are running on the final tcp instance rather than any temporary instance.

Using innodb_initialized as well to show of that more than one test is available, and the recovery time of innodb is a aspect that can be forgotten by those that don't frequently do mysql/mariadb admininstration.

tianon commented on mysql/Dockerfile.mariadb:
```diff
@@ -1,5 +1,5 @@
+ENV MARIADB_MYSQL_LOCALHOST_USER=1
 
-HEALTHCHECK CMD ["docker-healthcheck"]
+HEALTHCHECK CMD ["/usr/local/bin/healthcheck.sh", "--su=mysql", "--connect", "--innodb_initialized"]
```
One more minor nit -- I think we should let `PATH` make this easier to read: :smile:

```suggestion
HEALTHCHECK CMD ["healthcheck.sh", "--su=mysql", "--connect", "--innodb_initialized"]
```

grooverdan committed:
Update mysql/Dockerfile.mariadb

use implicit path

Co-authored-by: Tianon Gravi <admwiggin@gmail.com>

grooverdan commented:
Thanks @tianon
